Redo build system in cmake

jagerman opened the pull request.

jagerman committed:
Redo build system in cmake

despair86 commented:
why

notlesh commented:
> why

Because `qmake` is moderately shitty, and also because we're using `cmake` for `lokinet`

neuroscr reviewed (commented):
You missed the svg lib for mac

jagerman committed:
Update readme with qmake -> cmake

jagerman committed:
Build a .app with all the crap on macos

jagerman committed:
icns: no background, make cmake reconvert on script change

jagerman committed:
Put lokinet-gui.app building behind a cmake option

This lets someone build and use it locally just the same way they would
on Linux, but still allows building the bundle for producing an "app".

jagerman committed:
Tell macdeployqt to overwrite so you can make multiple times

jagerman committed:
Add some install paths

jagerman committed:
Change bundle name to CamelCase

jagerman committed:
Add codesigning option

jagerman committed:
Add --force to codesign

Otherwise codesign fails if it is already signed (even if something may
need signature updating).

jagerman committed:
Turn on hardening macos voodoo; verbose++

jagerman committed:
codesign automatically (when set) and be more verbose

jagerman committed:
Try -qmldir= to fix it

notlesh commented on README.md:
```diff
@@ -19,17 +19,29 @@ Then build and run:
+which will produce a lokinet-gui.app executable app bundle directory containing the binary,
+resources, and dependencies.
+
+Note that building this way additionally requires imagemagick installed and built with SVG support.
```
Apple also has a built in utility for creating a ICNS file/dir/whatever from a PNG, so we could avoid this dep by maintaining a PNG version of the icon. That's probably not worth it unless the `imagemagick` dep is problematic.

jagerman replied:
I think it's fine; imagemagick is pretty easily available.

notlesh commented on tools/mk-icns.sh:
```diff
@@ -0,0 +1,20 @@
+    cp "${outdir}/icon_${double}x${double}.png" "${outdir}/icon_${size}x${size}@2x.png"
+done
+
+iconutil -c icns "${outdir}"
```
Is there anything to clean up here?

jagerman replied:
I mean, it's a build directory...

notlesh replied:
Fair enough

notlesh commented on CMakeLists.txt:
```diff
@@ -0,0 +1,104 @@
+
+    if(MACOS_APP)
+        add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/lokinet-gui.icns
+            COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/tools/mk-icns.sh ${CMAKE_CURRENT_SOURCE_DIR}/res/images/icon.svg ${CMAKE_CURRENT_BINARY_DIR}/lokinet-gui.icns
```
Will this be invoked every time, or only when the `icns` file/dir/whatever needs to be generated?

jagerman replied:
The latter.

jagerman committed:
Try an entitlement

jagerman committed:
Try another entitlement
